# Data Analysis of all-time top bowlers in test matches

The greatest bowlers of all time has been listed in the dataset in separate rows according to total wickets they took in the span of their career as a cricketer. Along with wickets, their are other attributes that measures the performance of each player over their career. 

**Attributes/Columns:**
 - Player - name of the players
 - Span - starting and ending year of the players' career
 - Mat - matches played 
 - Inns - innings played 
 - Balls - balls bowled 
 - Runs - Runs conceded
 - Wkts - Wickets taken 
 - BBI - Best Bowling in Innings
 - BBM - Best Bowling in Match
 - Ave - Bowling Average
 - Econ - Economy Rate = Average number of runs conceded per over  
 - SR - Bowling Strike rate 
 - 5 - number of times the player took 5 wickets in an inning
 - 10 - number of times the player took 10 wickets in an inning

**Reference** 
<br>Data Source: https://stats.espncricinfo.com/ci/content/records/93276.html

### Importing packages

In [1]:
import numpy as np # linear Algebra
import pandas as pd # Data Processing

### Reading the dataset

In [2]:
# Naming the DataFrame - df
# Reading the .xlsx file using pandas: pd.read_csv("<location of dataset>")

df = pd.read_csv("wickets.csv")

# Displaying the first 10 rows of the DataFrame
display(df.head(10))

Player       Span   Mat  Inns  Balls   Runs  Wkts    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010   133   230  44039  18180   800   9/51   
1           SK Warne (AUS)  1992-2007   145   273  40705  17995   708   8/71   
2        JM Anderson (ENG)  2003-2021  164*   304  35079  16575   623   7/42   
3         A Kumble (INDIA)  1990-2008   132   236  40850  18355   619  10/74   
4         GD McGrath (AUS)  1993-2007   124   243  29248  12186   563   8/24   
5          SCJ Broad (ENG)  2007-2021   149   274  29863  14590   524   8/15   
6            CA Walsh (WI)  1984-2001   132   242  30019  12688   519   7/37   
7            DW Steyn (SA)  2004-2019    93   171  18608  10077   439   7/51   
8      N Kapil Dev (INDIA)  1978-1994   131   227  27740  12867   434   9/83   
9        HMRKB Herath (SL)  1999-2018    93   170  25993  12157   433  9/127   

      BBM    Ave  Econ    SR   5  10  
0  16/220  22.72  2.47  55.0  67  22  
1  12/128  25.41  2.65  57.4  37  10  
2   11/71  26.60  2.83  56.3  30   3  
3  14/149  29.65  2.69  65.9  35   8  
4   10/27  21.64  2.49  51.9  29   3  
5  11/121  27.84  2.93  56.9  18   3  
6   13/55  24.44  2.53  57.8  22   3  
7   11/60  22.95  3.24  42.3  26   5  
8  11/146  29.64  2.78  63.9  23   2  
9  14/184  28.07  2.80  60.0  34   9

### Number of rows and columns

In [3]:
# <name_of_DataFrame>.shape
df.shape

# output - (total number of rows, total number of columns)

(79, 14)

The DataFrame contains **79 rows** (players) and **14 columns** (attributes).

### Descriptive Statistics
<br>Output shows the following details of all numerical attributes in the dataset: 
- **count** (number of observations) 
- **mean** (average of all values)
- **std** (standard deviation)
- **min** (minimum value among all the observations)
- **25%** (the value at 25th percentile i.e. 25% of obversations/data has values less than the mentioned value)
- **50%** (median value or said the value at 50th percentile i.e. 50% of obversations/data has values less than the mentioned value)
- **75%** (the value at 75th percentile i.e. 75% of obversations/data has values less than the mentioned value)
- **max** (maximum value among all the observations)

In [4]:
# <name_of_DataFrame>.describe()
df.describe()

Inns         Balls          Runs        Wkts        Ave  \
count   79.000000     79.000000     79.000000   79.000000  79.000000   
mean   144.911392  18638.354430   8599.354430  317.215190  27.469747   
std     51.180222   7199.256972   3085.168807  121.924911   3.655658   
min     67.000000   8785.000000   4846.000000  200.000000  20.940000   
25%    110.000000  13583.000000   6456.500000  229.000000  24.500000   
50%    129.000000  16498.000000   7742.000000  266.000000  28.000000   
75%    169.000000  21742.500000   9756.000000  374.500000  29.870000   
max    304.000000  44039.000000  18355.000000  800.000000  34.790000   

            Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  
mean    2.806835  59.193671  16.354430   2.797468  
std     0.351577   9.350132   9.642372   3.235935  
min     1.980000  41.200000   3.000000   0.000000  
25%     2.600000  53.300000   9.500000   1.000000  
50%     2.820000  57.400000  14.000000   2.000000  
75%     3.080000  63.950000  20.500000   3.500000  
max     3.460000  91.900000  67.000000  22.000000

**Overview:**
1. The average number of  matches played by the top bowlers is 80, with minimum of 37 matches and maximum of 166.
2. The average number of runs made by the bowlers is 8595. While the highest runs is 18355, only 25% scored more than 9756 runs. 
4. The average number of wickets taken by the bowlers is 317, with 75% of bowlers scoring less than 375.

### Data Types and Missing Values

In [5]:
# <name_of_DataFrame>.info()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     object 
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(6), object(5)
memory usage: 8.8+ KB


**Overview:**
1. Total number of observations/players: 79
2. Total number of attributes/variables: 14
3. Total number of object (string/mixed) data type: 4 (Player, Span, BBI, BBM)
4. Total number of integer (positive/negative/zero) data type: 7 (Mat, Inns, Balls, Runs, Wkts, 5, 10)
5. Total number of float (floating point number) data type: 3 (Ave, Econ, SR)
6. No missing data

### Renaming columns

In [6]:
# <name_of_DataFrame> = <name_of_DataFrame>.rename(columns={'<name_of_column' : '<new_name>'})
df = df.rename(columns={'Mat':'Matches', 
                        'Inns':'Innings',
                        'Wkts':'Wickets',
                        'Ave': 'Average',
                        5: '5_wickets',
                        10:'10_wickets'})

list(df.columns) # or use print(df.columns)

['Player',
 'Span',
 'Matches',
 'Innings',
 'Balls',
 'Runs',
 'Wickets',
 'BBI',
 'BBM',
 'Average',
 'Econ',
 'SR',
 '5',
 '10']

### Removing columns

In [7]:
# <name_of_DataFrame>.drop('<name_of_column>', axis= )
# axis=1 and  is used to select columns and rows, respectively
# inplace=True used to edit the original DataFrame
df.drop("BBI", axis=1, inplace=True)
df.drop("BBM", axis=1, inplace=True)

print(df.columns) # or use list(df.columns)

Index(['Player', 'Span', 'Matches', 'Innings', 'Balls', 'Runs', 'Wickets',
       'Average', 'Econ', 'SR', '5', '10'],
      dtype='object')


### Splitting 'Player' Column and Creating New Columns 

In [8]:
# Removing the warning that one gets for the changes (as of August, 2021) in 'str.replace' syntax
import warnings
warnings.filterwarnings("ignore")

# Creating a new dataframe that stores information before and after '(' and removes '('
# A new Dataframe is created because otherwise rest of the columns are removed from the main DataFrame
df_player = df["Player"].str.split("(", expand=True)

# The columns (0,1) of new DataFrame is concatenated with the main Dataframe; axis=1 refers to column
df = pd.concat([df, df_player], axis=1)

# Removing the column 'Player'
df.drop("Player", axis=1, inplace=True)

# Renaming the column names of the newly concatenated DataFrame
df = df.rename(columns={0: "Player", 1: "Country"})

# Removing ")" from the 'Country' column
df["Country"] = df["Country"].str.replace(")", "")

display(df.head(3))

Span Matches  Innings  Balls   Runs  Wickets  Average  Econ    SR   5  \
0  1992-2010     133      230  44039  18180      800    22.72  2.47  55.0  67   
1  1992-2007     145      273  40705  17995      708    25.41  2.65  57.4  37   
2  2003-2021    164*      304  35079  16575      623    26.60  2.83  56.3  30   

   10           Player Country  
0  22  M Muralitharan   ICC/SL  
1  10        SK Warne      AUS  
2   3     JM Anderson      ENG

### Players who bowled at ICC

In [9]:
# Checking if player played at ICC using  function definition
def icc_check(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"

# Creating new column named 'Played_ICC' with 'Yes' or 'No'
df["Played_ICC"] = df["Country"].apply(icc_check)

display(df.head(3))

Span Matches  Innings  Balls   Runs  Wickets  Average  Econ    SR   5  \
0  1992-2010     133      230  44039  18180      800    22.72  2.47  55.0  67   
1  1992-2007     145      273  40705  17995      708    25.41  2.65  57.4  37   
2  2003-2021    164*      304  35079  16575      623    26.60  2.83  56.3  30   

   10           Player Country Played_ICC  
0  22  M Muralitharan   ICC/SL        Yes  
1  10        SK Warne      AUS         No  
2   3     JM Anderson      ENG         No

In [10]:
df["Played_ICC"].value_counts()

No     74
Yes     5
Name: Played_ICC, dtype: int64

There are **5 bowlers who played at ICC**. 

### Nationalities of the Bowlers

In [11]:
# Removing the string 'ICC/' from the 'Country' column
df["Country"] = df["Country"].str.replace("ICC/", "")
df["Country"] = df["Country"].str.replace("/ICC", "")
len(df["Country"].value_counts())

10

The 79 bowlers of the DataFrame are from **10 different countries**.

In [12]:
# Checking the number of players from each country
df["Country"].value_counts()

# df["Country"].str.contains("AUS").sum()
# df["Country"].str.contains("Bangladesh").sum()

AUS      18
ENG      15
INDIA    10
WI        9
SA        8
NZ        7
PAK       7
SL        3
ZIM       1
BDESH     1
Name: Country, dtype: int64

Among the 11 nationalities, **18 bowlers are from Australia** and only **1 is from Bangladesh**.

### Splitting 'Span' Column and Creating New Columns 

In [13]:
# Creating new column "Start_year" with values of index 0 to 3, in "Span"
df["Start_year"] = df["Span"].str[0:4]
# Creating new column "End_year" with values of index 5 to end, in "Span" (4 index is the "-")
df["End_year"] = df["Span"].str[5:]

# Removing the column "Span" 
df = df.drop("Span", axis=1)

# Changing the datatype because the columns 'Start_year' and 'End_year' were String object
df["Start_year"] = df["Start_year"].astype("int")
df["End_year"] = df["End_year"].astype("int")

# Creating new column that includes the difference of 'End_year' and 'Start_year'
df["Span"] = df["End_year"] - df["Start_year"]

# Removing the column "End_year" and "Start_year" 
df = df.drop("Start_year", axis = 1)
df = df.drop("End_year", axis = 1)

display(df.head(3))

# Another code same as the Spliting 'Player' column code:
# df_Span = df["Span"].str.split("-", expand=True)
# df = pd.concat([df, df_Span], axis=1)
# df = df.rename(columns={0: "start_year", 1: "end_year"})

Matches  Innings  Balls   Runs  Wickets  Average  Econ    SR   5  10  \
0     133      230  44039  18180      800    22.72  2.47  55.0  67  22   
1     145      273  40705  17995      708    25.41  2.65  57.4  37  10   
2    164*      304  35079  16575      623    26.60  2.83  56.3  30   3   

            Player Country Played_ICC  Span  
0  M Muralitharan       SL        Yes    18  
1        SK Warne      AUS         No    15  
2     JM Anderson      ENG         No    18

### Rearranging Columns

In [14]:
# Rearranging the order of the columns by adding it to a new DataFrame
col_sequence = ['Player', 'Country', 'Span', 'Matches', 'Innings', 'Balls', 'Runs', 'Wickets',
                    'Average', 'Econ', 'SR', '5', '10']

# Changing the DataFrame
df = df[col_sequence]

print(df.columns)

Index(['Player', 'Country', 'Span', 'Matches', 'Innings', 'Balls', 'Runs',
       'Wickets', 'Average', 'Econ', 'SR', '5', '10'],
      dtype='object')


### Bowlers with Longest and Shortest Career

In [15]:
# <name_of_DataFrame>.sort_values(by=<name_of_column>, ascending = False)
# ascending=True, for organizing in ascending order (Set by default)
# ascending=False, for organizing in descending order

# Organizing the "Span" column in descending order
df.sort_values(by = "Span", ascending = False)

Player Country  Span Matches  Innings  Balls   Runs  Wickets  \
21      Imran Khan      PAK    21      88      142  19458   8258      362   
55       GS Sobers       WI    20      93      159  21599   7999      235   
9     HMRKB Herath       SL    19      93      170  25993  12157      433   
0   M Muralitharan       SL    18     133      230  44039  18180      800   
2      JM Anderson      ENG    18    164*      304  35079  16575      623   
..              ...     ...   ...     ...      ...    ...    ...      ...   
54      Yasir Shah      PAK     7     46*       84  13607   7248      235   
61     SJ Harmison      ENG     7      63      115  13375   7192      226   
72    JR Hazlewood      AUS     7      55      103  11887   5438      212   
71        K Rabada       SA     6      47       86   8785   4846      213   
44        GP Swann      ENG     5      60      109  15349   7642      255   

    Average  Econ    SR   5  10  
21    22.81  2.54  53.7  23   6  
55    34.03  2.22  91.9   6   0  
9     28.07  2.80  60.0  34   9  
0     22.72  2.47  55.0  67  22  
2     26.60  2.83  56.3  30   3  
..      ...   ...   ...  ..  ..  
54    30.84  3.19  57.9  16   3  
61    31.82  3.22  59.1   8   1  
72    25.65  2.74  56.0   9   0  
71    22.75  3.30  41.2  10   4  
44    29.96  2.98  60.1  17   3  

[79 rows x 13 columns]

**5 Bowlers with Longest Career:**
1. Imran Khan - 21 years
2. GS Sobers - 20 years
3. HMRKB Herath - 19 years
4. M Muralitharan - 18 years
5. JM Anderson - 18 years

**5 Bowlers with Shortest Career:**
1. Yasir Shah - 7 years
2. SJ Harmison - 7 years
3. JR Hazlewood - 7 years
4. K Rabada - 6 years
5. GP Swann - 5 years

### Bowler with lowest Economy Rate

In [16]:
df.sort_values(by = "Econ").head(1)

Player Country  Span Matches  Innings  Balls  Runs  Wickets  Average  \
32  LR Gibbs       WI    18      79      148  27115  8989      309    29.09   

    Econ    SR   5  10  
32  1.98  87.7  18   2

The Bowler with lowest Economy Rate is **LR Gibbs** (1.98)

### Bowler with lowest Strike Rate

In [17]:
df.sort_values(by = "SR").head(1)

Player Country  Span Matches  Innings  Balls  Runs  Wickets  Average  \
71  K Rabada       SA     6      47       86   8785  4846      213    22.75   

    Econ    SR   5  10  
71   3.3  41.2  10   4

The Bowler with lowest Strike Rate is **K Rabada** (41.2).

### Bowler with lowest Average

In [18]:
df.sort_values(by = "Average").head(1)

# The formula for calculating Bowling Average
# df['avg'] = df['Runs']/df['Wickets']

Player Country  Span Matches  Innings  Balls  Runs  Wickets  \
19  MD Marshall       WI    13      81      151  17584  7876      376   

    Average  Econ    SR   5  10  
19    20.94  2.68  46.7  22   4

The Bowler with lowest Average is **MD Marshal** (20.94).